src/critics/path_align_critic.cpp

```diff
@@ -58,46 +61,68 @@ void PathAlignCritic::score(models::CriticFunctionData & data)
       double dx = P(0) - P3(t, p, 0);
       double dy = P(1) - P3(t, p, 1);
       return std::hypot(dx, dy);
     };
 
-  size_t max_s = 0;
-  for (size_t t = 0; t < trajectories_count; ++t) {
-    double mean_dist = 0;
-    for (size_t p = 0; p < trajectories_points_count; p += trajectory_point_step_) {
-      double min_dist = std::numeric_limits<double>::max();
-      size_t min_s = 0;
-      for (size_t s = 0; s < reference_segments_count; s += path_point_step_) {
-        xt::xtensor_fixed<double, xt::xshape<2>> P;
-        if (segment_short(s)) {
-          P[0] = P1(s, 0);
-          P[1] = P1(s, 1);
-        } else if (double u = evaluate_u(t, p, s); u <= 0) {
-          P[0] = P1(s, 0);
-          P[1] = P1(s, 1);
-        } else if (u >= 1) {
-          P[0] = P2(s, 0);
-          P[1] = P2(s, 1);
-        } else {
-          P[0] = P1(s, 0) + u * P2_P1_diff(s, 0);
-          P[1] = P1(s, 1) + u * P2_P1_diff(s, 1);
-        }
-        auto dist = evaluate_dist(std::move(P), t, p);
-        if (dist < min_dist) {
-          min_s = s;
-          min_dist = dist;
-        }
-      }
-      max_s = std::max(max_s, min_s);
-      mean_dist += min_dist;
-    }
 
-    cost(t) = mean_dist / trajectories_points_count;
+  xt::xtensor<double, 1> trajectories_lengths;
+  {
+    auto next = xt::view(P3, xt::all(), xt::range(1, _), xt::range(0, 2));
+    auto prev = xt::view(P3, xt::all(), xt::range(_, -1), xt::range(0, 2));
+    auto dist = xt::norm_sq(next - prev, {2});
+    trajectories_lengths = xt::sum(dist, {1});
   }
+  auto accumulated_path_distances = xt::cumsum(P2_P1_norm_sq);
+
+  std::atomic<size_t> max_s = 0;
+
+  std::vector<size_t> trajectories_idxs(trajectories_count);
+  std::iota(trajectories_idxs.begin(), trajectories_idxs.end(), 0);
+
+  std::for_each(
+    std::execution::par_unseq, trajectories_idxs.begin(), trajectories_idxs.end(),
+    [&](int t) {
+      double mean_dist = 0;
+      double trajectory_len = trajectories_len(t);
+      for (size_t p = 0; p < trajectories_points_count; p += trajectory_point_step_) {
+        double min_dist = std::numeric_limits<double>::max();
+        size_t min_s = 0;
+
+        for (size_t s = 0; s < reference_segments_count; s += path_point_step_) {
+          xt::xtensor_fixed<double, xt::xshape<2>> P;
+          if (segment_short(s)) {
+            P[0] = P1(s, 0);
+            P[1] = P1(s, 1);
+          } else if (double u = evaluate_u(t, p, s); u <= 0) {
+            P[0] = P1(s, 0);
+            P[1] = P1(s, 1);
+          } else if (u >= 1) {
+            P[0] = P2(s, 0);
+            P[1] = P2(s, 1);
+          } else {
+            P[0] = P1(s, 0) + u * P2_P1_diff(s, 0);
+            P[1] = P1(s, 1) + u * P2_P1_diff(s, 1);
+          }
+          auto dist = evaluate_dist(P, t, p);
+          if (dist < min_dist) {
+            min_s = s;
+            min_dist = dist;
+          }
+
+          if (accumulated_path_distances(s) > trajectories_lengths(t)) {
```

Comment:
I think `if(accumulated_path_distances(s) > trajectories_lengths(t))` we should set a boolean flag to do 1 more iteration since the closest could be up to 1 past the same length and then break if that flag is true

Reply:
Do we still need comment after the bool addition? It seems self-documenting